File: recyclerview-lib/src/main/java/net/maxsmr/android/recyclerview/adapters/BaseRecyclerViewAdapter.java
package net.maxsmr.android.recyclerview.adapters;

import android.content.Context;
import android.database.Observable;
import android.support.annotation.CallSuper;
import android.support.annotation.LayoutRes;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;

@SuppressWarnings({"WeakerAccess", "unused"})
public abstract class BaseRecyclerViewAdapter<I, VH extends BaseRecyclerViewAdapter.ViewHolder> extends RecyclerView.Adapter<VH> {

    @NotNull
    protected final ItemsEventsObservable<I, VH> mItemsEventsObservable = new ItemsEventsObservable<>();

    @NotNull
    protected final Context mContext;

    @LayoutRes
    protected final int mBaseItemLayoutId;

    @NotNull
    protected final ArrayList<I> mItems = new ArrayList<>();

    protected boolean mNotifyOnChange = true;

    protected BaseRecyclerViewAdapter(@NotNull Context context) {
        this(context, 0, null);
    }

    protected BaseRecyclerViewAdapter(@NotNull Context context, @LayoutRes int baseItemLayoutId, @Nullable Collection<I> items) {
        this.mContext = context;
        this.mBaseItemLayoutId = baseItemLayoutId;
        this.setItems(items);
    }

    public void registerItemsEventsListener(@NotNull ItemsEventsListener<I> listener) {
        mItemsEventsObservable.registerObserver(listener);
    }

    public void unregisterItemsEventsListener(@NotNull ItemsEventsListener<I> listener) {
        mItemsEventsObservable.unregisterObserver(listener);
    }

    protected final void rangeCheck(int position) {
        synchronized (mItems) {
            if (position < 0 || position >= mItems.size()) {
                throw new IndexOutOfBoundsException("incorrect position: " + position);
            }
        }
    }

    protected final void rangeCheckForAdd(int position) {
        synchronized (mItems) {
            if (position < 0 || position > mItems.size()) {
                throw new IndexOutOfBoundsException("incorrect add position: " + position);
            }
        }
    }

    @NotNull
    public final ArrayList<I> getItems() {
        synchronized (mItems) {
            return new ArrayList<>(mItems);
        }
    }

    @Nullable
    public final I getFirstItem() {
        if (!isEmpty()) {
            return getItem(0);
        }
        return null;
    }

    @Nullable
    public final I getLastItem() {
        if (!isEmpty()) {
            return getItem(getItemCount() - 1);
        }
        return null;
    }

    @Nullable
    public final I getItem(int at) throws IndexOutOfBoundsException {
        synchronized (mItems) {
            rangeCheck(at);
            return mItems.get(at);
        }
    }

    public final int indexOf(I item) {
        synchronized (mItems) {
            return mItems.indexOf(item);
        }
    }

    public final int lastIndexOf(I item) {
        synchronized (mItems) {
            return mItems.lastIndexOf(item);
        }
    }

    public void sort(@NotNull Comparator<? super I> comparator) {
        synchronized (mItems) {
            Collections.sort(mItems, comparator);
            if (mNotifyOnChange) {
                notifyDataSetChanged();
            }
        }
    }

    /**
     * @param items null for reset adapter
     */
    public final void setItems(@Nullable Collection<I> items) {
        synchronized (mItems) {
            if (!mItems.equals(items)) {
                clearItems();
                if (items != null) {
                    this.mItems.addAll(items);
                }
                onItemsSet();
            }
        }
    }

    protected void onItemsSet() {
        if (mNotifyOnChange) {
            notifyDataSetChanged();
        }
        mItemsEventsObservable.notifyItemsSet(getItems());
    }

    public final void clearItems() {
        synchronized (mItems) {
            if (!isEmpty()) {
                int previousSize = getItemCount();
                mItems.clear();
                onItemsRangeRemoved(0, previousSize - 1, previousSize);
            }
        }
    }

    protected void onItemsCleared(int previousSize) {
    }

    public final void addItem(int to, @Nullable I item) throws IndexOutOfBoundsException {
        synchronized (mItems) {
            rangeCheckForAdd(to);
            mItems.add(to, item);
            onItemAdded(to, item);
        }
    }

    public final void addItem(@Nullable I item) {
        addItem(getItemCount(), item);
    }

    public final void addFirstItem(@Nullable I item) throws IndexOutOfBoundsException {
        addItem(0, item);
    }

    @CallSuper
    protected void onItemAdded(int to, @Nullable I item) {
        if (mNotifyOnChange) {
            if (to == 0) {
                notifyDataSetChanged();
            } else {
                notifyItemInserted(to);
            }
        }
        mItemsEventsObservable.notifyItemAdded(to, item);

    }

    public final void addItems(int to, @Nullable Collection<I> items) {
        synchronized (mItems) {
            rangeCheckForAdd(to);
            if (items != null) {
                mItems.addAll(to, items);
                onItemsAdded(to, items);
            }
        }
    }

    public final void addItems(@Nullable Collection<I> items) {
        addItems(getItemCount(), items);
    }

    @CallSuper
    protected void onItemsAdded(int to, @NotNull Collection<I> items) {
        if (mNotifyOnChange) {
            if (to == 0) {
                notifyDataSetChanged();
            } else {
                notifyItemRangeInserted(to, items.size());
            }
        }
        mItemsEventsObservable.notifyItemsAdded(to, items);

    }

    public final void setItem(int in, @Nullable I item) {
        synchronized (mItems) {
            rangeCheck(in);
            mItems.set(in, item);
            onItemSet(in, item);
        }
    }

    @CallSuper
    protected void onItemSet(int in, @Nullable I item) {
        if (mNotifyOnChange) {
            notifyItemChanged(in);
        }
        mItemsEventsObservable.notifyItemSet(in, item);
    }

    @Nullable
    public final I replaceItem(int in, @Nullable I newItem) {
        synchronized (mItems) {
            rangeCheck(in);
            setNotifyOnChange(false);
            I replacedItem = getItem(in);
            mItems.remove(in);
            onItemRemoved(in, replacedItem);
            addItem(in, newItem);
            setNotifyOnChange(true);
            notifyItemChanged(in);
            return replacedItem;
        }
    }

    @Nullable
    public final I replaceItem(@Nullable I replaceableItem, @Nullable I newItem) {
        return replaceItem(indexOf(replaceableItem), newItem);
    }

    @NotNull
    public final List<I> replaceItemsRange(int from, int to, @Nullable Collection<I> newItems) {
        synchronized (mItems) {
            setNotifyOnChange(false);
            List<I> replacedItems = removeItemsRange(from, to);
            addItems(newItems);
            setNotifyOnChange(true);
            notifyItemRangeChanged(from, to - from);
            return replacedItems;
        }
    }

    @Nullable
    public final I removeItem(@Nullable I item) {
        return removeItem(indexOf(item));
    }

    @Nullable
    public final I removeItem(int from) {
        synchronized (mItems) {
            rangeCheck(from);
            I removedItem = getItem(from);
            mItems.remove(from);
            onItemRemoved(from, removedItem);
            return removedItem;
        }
    }

    @CallSuper
    protected void onItemRemoved(int from, @Nullable I item) {
        if (mNotifyOnChange) {
            notifyItemRemoved(from);
        }
        mItemsEventsObservable.notifyItemRemoved(from, item);
    }

    @NotNull
    public final List<I> removeItemsRange(int from, int to) {
        rangeCheck(from);
        rangeCheck(to);
        synchronized (mItems) {
            int previousSize = getItemCount();
            List<I> removed = new ArrayList<>();
            int position = 0;
            Iterator<I> iterator = mItems.iterator();
            while (iterator.hasNext()) {
                if (position >= from && position <= to) {
                    I item = iterator.next();
                    iterator.remove();
                    removed.add(item);
                }
                position++;
            }
            if (!removed.isEmpty()) {
                onItemsRangeRemoved(from, to, previousSize);
            }
            return removed;
        }
    }

    public final void removeAllItems() {
        synchronized (mItems) {
            if (!isEmpty()) {
                removeItemsRange(0, getItemCount() - 1);
            }
        }
    }

    @CallSuper
    protected void onItemsRangeRemoved(int from, int to, int previousSize) {
        if (mNotifyOnChange) {
            notifyItemRangeRemoved(from, to - from);
        }
        mItemsEventsObservable.notifyItemsRangeRemoved(from, to, previousSize);
        if (from == 0 && to == previousSize - 1) {
            onItemsCleared(previousSize);
        }
    }

    protected final View onInflateView(ViewGroup parent, int viewType) {
        return LayoutInflater.from(parent.getContext())
                .inflate(getLayoutIdForViewType(viewType), parent, false);
    }

    @Nullable
    protected View getClickableView(@NotNull VH holder) {
        return holder.itemView;
    }

    @LayoutRes
    protected int getLayoutIdForViewType(int viewType) {
        return mBaseItemLayoutId;
    }

    @Override
    @NotNull
    public abstract VH onCreateViewHolder(@NotNull ViewGroup parent, int viewType);

    @Override
    public final void onBindViewHolder(@NotNull VH holder, int position) {
        final I item = (position >= 0 && position < mItems.size()) ? mItems.get(position) : null;
        processItem(holder, item, position);
    }

    protected boolean allowSetClickListener(@Nullable final I item, final int position) {
        return !isItemEmpty(item, position);
    }

    protected boolean allowSetLongClickListener(@Nullable final I item, final int position) {
        return allowSetClickListener(item, position);
    }

    protected boolean allowFillHolderForItem(@NotNull VH holder, @Nullable final I item, final int position) {
        return true;
    }

    protected boolean isItemEmpty(@Nullable final I item, final int position) {
        return item == null;
    }

    @SuppressWarnings("unchecked")
    @CallSuper
    protected void processItem(@NotNull VH holder, @Nullable final I item, final int position) {

        if (!isItemEmpty(item, position) && item != null) {

            View clickableView = getClickableView(holder);

            if (clickableView != null) {
                clickableView.setOnClickListener(v -> {
                    if (allowSetClickListener(item, position)) {
                        mItemsEventsObservable.notifyItemClick(position, item);
                    }
                });
                clickableView.setOnLongClickListener(v -> {
                    if (allowSetLongClickListener(item, position)) {
                        return mItemsEventsObservable.notifyItemLongClick(position, item);
                    }
                    return false;
                });
            }

            if (allowFillHolderForItem(holder, item, position)) {
                holder.displayData(position, item, getItemCount());
            }

        } else {

            if (allowFillHolderForItem(holder, item, position)) {
                holder.displayEmptyData(position, item, getItemCount());
            }
        }
    }

    public boolean isEmpty() {
        return getItemCount() == 0;
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    public final boolean isNotifyOnChange() {
        return mNotifyOnChange;
    }

    public void setNotifyOnChange(boolean enable) {
        mNotifyOnChange = enable;
    }

    @Override
    @CallSuper
    public void onViewRecycled(@NotNull VH holder) {
        super.onViewRecycled(holder);
        holder.onViewRecycled();
        holder.itemView.setOnClickListener(null);
    }

    @CallSuper
    public void release() {
        mItemsEventsObservable.unregisterAll();
    }

    public static abstract class ViewHolder<I> extends RecyclerView.ViewHolder {

        @NotNull
        protected final Context context;

        public ViewHolder(@NotNull ViewGroup parent, @LayoutRes int layoutId) {
            this(LayoutInflater.from(parent.getContext()).inflate(layoutId, parent, false));
        }

        public ViewHolder(@NotNull View view) {
            super(view);
            this.context = view.getContext();
        }

        public void displayData(int position, @NotNull final I item, int count) {
            itemView.setVisibility(View.VISIBLE);
        }

        public void displayEmptyData(int position, @Nullable final I item, int count) {
            itemView.setVisibility(View.GONE);
        }

        protected void onViewRecycled() {
        }
    }

    protected static class ItemsEventsObservable<I, VH extends BaseRecyclerViewAdapter.ViewHolder>
            extends Observable<ItemsEventsListener<I>> {

        protected void notifyItemClick(int position, I item) {
            synchronized (mObservers) {
                for (ItemsEventsListener<I> l : mObservers) {
                    l.onItemClick(position, item);
                }
            }
        }

        protected boolean notifyItemLongClick(int position, I item) {
            synchronized (mObservers) {
                boolean consumed = false;
                for (ItemsEventsListener<I> l : mObservers) {
                    if (l.onItemLongClick(position, item)) {
                        consumed = true;
                    }
                }
                return consumed;
            }
        }

        protected void notifyItemAdded(int to, @Nullable I item) {
            synchronized (mObservers) {
                for (ItemsEventsListener<I> l : mObservers) {
                    l.onItemAdded(to, item);
                }
            }
        }

        protected void notifyItemsAdded(int to, @NotNull Collection<I> items) {
            synchronized (mObservers) {
                for (ItemsEventsListener<I> l : mObservers) {
                    l.onItemsAdded(to, items);
                }
            }
        }

        protected void notifyItemSet(int to, I item) {
            synchronized (mObservers) {
                for (ItemsEventsListener<I> l : mObservers) {
                    l.onItemSet(to, item);
                }
            }
        }

        protected void notifyItemsSet(@NotNull List<I> items) {
            synchronized (mObservers) {
                for (ItemsEventsListener<I> l : mObservers) {
                    l.onItemsSet(items);
                }
            }
        }

        protected void notifyItemRemoved(int from, I item) {
            synchronized (mObservers) {
                for (ItemsEventsListener<I> l : mObservers) {
                    l.onItemRemoved(from, item);
                }
            }
        }

        protected void notifyItemsRangeRemoved(int from, int to, int previousSize) {
            synchronized (mObservers) {
                for (ItemsEventsListener<I> l : mObservers) {
                    l.onItemsRangeRemoved(from, to, previousSize);
                }
            }
        }
    }

    public interface ItemsEventsListener<I> {

        void onItemClick(int position, I item);

        /**
         * @return true if event consumed
         */
        boolean onItemLongClick(int position, I item);

        void onItemAdded(int to, @Nullable I item);

        void onItemsAdded(int to, @NotNull Collection<I> items);

        void onItemSet(int to, I item);

        void onItemsSet(@NotNull List<I> items);

        void onItemRemoved(int from, I item);

        void onItemsRangeRemoved(int from, int to, int previousSize);
    }


}

File: recyclerview-lib/src/main/kotlin/net/maxsmr/android/recyclerview/adapters/BaseRecyclerViewAdapter.kt
package net.maxsmr.android.recyclerview.adapters

import android.content.Context
import android.database.Observable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.CallSuper
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.NO_ID
import net.maxsmr.android.recyclerview.adapters.diff.AutoNotifyDiffCallback
import net.maxsmr.android.recyclerview.adapters.diff.ItemInfo
import java.util.*

private const val INFINITE_SCROLL_LOOPS_COUNT = 100

abstract class BaseRecyclerViewAdapter<I, VH : BaseRecyclerViewAdapter.ViewHolder<*>>(
        protected val context: Context,
        @LayoutRes
        protected val baseItemLayoutId: Int = 0,
        items: Collection<I>? = null
) : RecyclerView.Adapter<VH>() {

    val firstItem: I?
        get() = if (isNotEmpty) {
            getItem(0)
        } else null

    val lastItem: I?
        get() {
            return if (isNotEmpty) {
                getItem(itemCount - 1)
            } else null
        }

    val isEmpty: Boolean
        get() = itemCount == 0

    val isNotEmpty: Boolean
        get() = !isEmpty

    val items = mutableListOf<I?>()
//        get() = field.toMutableList()

    protected val itemsEventsObservable = ItemsEventsObservable<I, VH>()

    /**
     * Маппинг позиция в адаптере <-> LayoutRes разметки в этой позиции
     */
    var viewTypeLayoutResMap = mapOf<Int, Int>()

    /**
     * Разрешить обновление элементов в адаптере
     * при изменении данных
     */
    var allowNotifyOnChange = true

    /**
     * Разрешить выборочное обновление элементов в адаптере
     * по позициям с изменившимися данными
     */
    var allowDiffNotifyOnChange = true

    var allowInfiniteScroll: Boolean = false

    protected var pendingFocusPosition = RecyclerView.NO_POSITION

    private var lastItemsInfo = listOf<ItemInfo>()

    init {
        setItems(items, false)
    }

    abstract override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH

    override fun getItemCount() = if (allowInfiniteScroll) INFINITE_SCROLL_LOOPS_COUNT * items.size else items.size;

    @CallSuper
    override fun onViewRecycled(holder: VH) {
        super.onViewRecycled(holder)
        holder.onViewRecycled()
    }

    override fun onBindViewHolder(holder: VH, position: Int) {
        val targetPosition = getListPosition(position)
        val item = if (targetPosition in 0 until items.size) items[targetPosition] else null
        bindItem(holder, item, targetPosition)
    }

    /**
     * @see RecyclerView.Adapter.getItemId
     */
    override fun getItemId(position: Int): Long =
            getItemStringId(position).hashCode().toLong()

    /**
     * Get the unique id from item at certain position
     *
     * @param position position of item
     * @return unique item id
     */
    open fun getItemStringId(position: Int): String = NO_ID.toString()

    /**
     * Get the item's hashcode at certain position
     *
     * @param position position of item
     * @return item's hashcode
     */
    open fun getItemHash(position: Int): String {
        val item = items[getListPosition(position)]
        return (item?.hashCode() ?: 0).toString()
    }

    @CallSuper
    open fun release() {
        itemsEventsObservable.unregisterAll()
        items.clear()
    }

    fun registerItemsEventsListener(listener: ItemsEventsListener<I>) {
        itemsEventsObservable.registerObserver(listener)
    }

    fun unregisterItemsEventsListener(listener: ItemsEventsListener<I>) {
        itemsEventsObservable.unregisterObserver(listener)
    }


    @Throws(IndexOutOfBoundsException::class)
    fun getItem(at: Int): I? {
        synchronized(items) {
            rangeCheck(at)
            return items[at]
        }
    }

    fun indexOf(item: I?): Int =
            synchronized(items) {
            return items.indexOf(item)
        }

    fun lastIndexOf(item: I): Int =
            synchronized(items) {
            items.lastIndexOf(item)
        }


    /**
     * Запросить фокус в указанном [index];
     * если не входит в текущий диапазон,
     * то только присвоить для дальнейших bind
     */
    fun requestFocus(index: Int) {
        pendingFocusPosition = index
        if (allowNotifyOnChange) {
            if (index in 0 until itemCount) {
                notifyItemChanged(index)
            }
        }
    }

    /**
     * @param items null for reset adapter
     */
    fun setItems(items: Collection<I>?, shouldNotify: Boolean = true) {
        if (shouldSetItems(items)) {
            clearItems()
            if (items != null) {
                this.items.addAll(items)
            }
            if (shouldNotify) {
                onItemsSet()
            }
        }
    }

    fun clearItems() {
        if (isNotEmpty) {
            val previousSize = itemCount
            items.clear()
            onItemsRangeRemoved(0, previousSize - 1, previousSize)
        }
    }

    @Throws(IndexOutOfBoundsException::class)
    fun addItem(to: Int, item: I?) {
        synchronized(items) {
            rangeCheckForAdd(to)
            items.add(to, item)
            onItemAdded(to, item)
        }
    }

    fun addItem(item: I?) {
        addItem(itemCount, item)
    }

    @Throws(IndexOutOfBoundsException::class)
    fun addFirstItem(item: I?) {
        addItem(0, item)
    }

    fun addItems(to: Int, items: Collection<I>?) {
        synchronized(this.items) {
            rangeCheckForAdd(to)
            if (items != null) {
                this.items.addAll(to, items)
                onItemsAdded(to, items)
            }
        }
    }

    fun addItems(items: Collection<I>?) {
        addItems(itemCount, items)
    }

    fun setItem(`in`: Int, item: I?) {
        rangeCheck(`in`)
        items[`in`] = item
        onItemSet(`in`, item)
    }

    fun replaceItem(`in`: Int, newItem: I?): I? {
        synchronized(items) {
            rangeCheck(`in`)
            allowNotifyOnChange = false
            val replacedItem = getItem(`in`)
            items.removeAt(`in`)
            onItemRemoved(`in`, replacedItem)
            addItem(`in`, newItem)
            allowNotifyOnChange = true
            notifyItemChanged(`in`)
            return replacedItem
        }
    }

    fun replaceItem(replaceableItem: I?, newItem: I?): I? {
        return replaceItem(indexOf(replaceableItem), newItem)
    }

    fun replaceItemsRange(from: Int, to: Int, newItems: Collection<I>?): List<I?> {
        allowNotifyOnChange = false
        val replacedItems = removeItemsRange(from, to)
        addItems(newItems)
        allowNotifyOnChange = true
        notifyItemRangeChanged(from, to - from)
        return replacedItems
    }

    fun removeItem(item: I?): I? {
        return removeItem(indexOf(item))
    }

    fun removeItem(from: Int): I? {
        rangeCheck(from)
        val removedItem = getItem(from)
        items.removeAt(from)
        onItemRemoved(from, removedItem)
        return removedItem
    }

    fun removeItemsRange(from: Int, to: Int): List<I?> {
        rangeCheck(from)
        rangeCheck(to)
        val previousSize = itemCount
        val removed = mutableListOf<I?>()
        var position = 0
        val iterator = items.iterator()
        while (iterator.hasNext()) {
            if (position in from..to) {
                val item = iterator.next()
                iterator.remove()
                removed.add(item)
            }
            position++
        }
        if (removed.isNotEmpty()) {
            onItemsRangeRemoved(from, to, previousSize)
        }
        return removed
    }

    fun removeAllItems() {
        if (isNotEmpty) {
            removeItemsRange(0, itemCount - 1)
        }
    }

    protected open fun areItemsEqual(items: Collection<I>?) = this.items == items

    // override to disable equals check on items set
    protected open fun shouldSetItems(items: Collection<I>?): Boolean = !areItemsEqual(items)

    protected open fun onInflateView(parent: ViewGroup, viewType: Int): View =
            LayoutInflater.from(parent.context)
                    .inflate(getLayoutIdForViewType(viewType), parent, false)

    protected open fun getClickableView(holder: VH): View? =
            holder.itemView

    protected open fun getLongClickableView(holder: VH): View? =
            getClickableView(holder)

    protected open fun getFocusableView(holder: VH): View? =
            getClickableView(holder)

    @LayoutRes
    protected open fun getLayoutIdForViewType(viewType: Int): Int =
            viewTypeLayoutResMap[viewType] ?: baseItemLayoutId

    @CallSuper
    protected open fun onItemSet(`in`: Int, item: I?) {
        itemsEventsObservable.notifyItemSet(`in`, item)
        if (allowNotifyOnChange) {
            notifyItemChanged(`in`)
        }
    }

    @CallSuper
    protected open fun onItemsAdded(to: Int, items: Collection<I>) {
        itemsEventsObservable.notifyItemsAdded(to, items)
        if (allowNotifyOnChange) {
            if (to == 0) {
                notifyDataSetChanged()
            } else {
                notifyItemRangeInserted(to, items.size)
            }
        }
    }

    @CallSuper
    protected open fun onItemAdded(to: Int, item: I?) {
        itemsEventsObservable.notifyItemAdded(to, item)
        if (allowNotifyOnChange) {
            if (to == 0) {
                notifyDataSetChanged()
            } else {
                notifyItemInserted(to)
            }
        }
    }

    @CallSuper
    protected open fun onItemsRangeRemoved(from: Int, to: Int, previousSize: Int) {
        itemsEventsObservable.notifyItemsRangeRemoved(from, to, previousSize)
        if (allowNotifyOnChange) {
            notifyItemRangeRemoved(from, to - from)
        }
        if (from == 0 && to == previousSize - 1) {
            onItemsCleared(previousSize)
        }
    }

    @CallSuper
    protected open fun onItemRemoved(from: Int, item: I?) {
        itemsEventsObservable.notifyItemRemoved(from, item)
        if (allowNotifyOnChange) {
            notifyItemRemoved(from)
        }
    }

    protected open fun onItemsCleared(previousSize: Int) {
        // override if needed
    }

    protected open fun onItemsSet() {
        itemsEventsObservable.notifyItemsSet(items)
        if (allowNotifyOnChange) {
            if (allowDiffNotifyOnChange) {
                diffNotifyDataSetChanged()
            } else {
                notifyDataSetChanged()
            }
        }
    }

    protected fun rangeCheck(position: Int) {
        if (position < 0 || position >= items.size) {
            throw IndexOutOfBoundsException("Incorrect position: $position")
        }
    }

    protected fun rangeCheckForAdd(position: Int) {
        if (position < 0 || position > items.size) {
            throw IndexOutOfBoundsException("Incorrect add position: $position")
        }
    }

    protected open fun allowSetClickListener(item: I?, position: Int) = !isItemEmpty(item, position)

    protected open fun allowSetLongClickListener(item: I?, position: Int) = allowSetClickListener(item, position)

    protected open fun allowFillHolderForItem(holder: VH, item: I?, position: Int) = true

    protected open fun allowBindFocusForItem(holder: VH, item: I?, position: Int) = pendingFocusPosition == position

    protected open fun isItemEmpty(item: I?, position: Int): Boolean = item == null

    @Suppress("UNCHECKED_CAST")
    @CallSuper
    protected open fun bindItem(holder: VH, item: I?, position: Int) {

        holder as ViewHolder<Any>

        val listPosition = getListPosition(position)
        val allowFill = allowFillHolderForItem(holder, item, listPosition)

        if (!isItemEmpty(item, listPosition) && item != null) {

            if (allowSetClickListener(item, listPosition)) {
                getClickableView(holder)?.setOnClickListener {
                    itemsEventsObservable.notifyItemClick(getListPosition(holder.adapterPosition), item)
                }
            }

            if (allowSetLongClickListener(item, listPosition)) {
                getLongClickableView(holder)?.setOnLongClickListener { itemsEventsObservable.notifyItemLongClick(holder.adapterPosition, item) }
            }

            if (allowFill) {
                holder.bindData(position, item, itemCount)
            }

        } else {

            if (allowFill) {
                holder.bindEmptyData(listPosition, item, itemCount)
            }
        }

        if (allowBindFocusForItem(holder, item, listPosition)) {
            bindItemFocus(holder, item, listPosition)
            itemsEventsObservable.notifyItemFocused(listPosition, item)
            pendingFocusPosition = RecyclerView.NO_POSITION
        }
    }

    /**
     * Реализация запроса фокуса на фокусабельной view
     */
    @CallSuper
    protected open fun bindItemFocus(holder: VH, item: I?, position: Int) {
        getFocusableView(holder)?.requestFocus()
    }

    /**
     * @return позиция с учётом включённости бесконечного скролла
     */
    protected fun getListPosition(adapterPosition: Int) = if (allowInfiniteScroll) {
        adapterPosition % items.size
    } else {
        adapterPosition
    }

    /**
     * Extract real items info, despite of infinite or ordinary scroll.
     */
    private fun extractRealItemInfo(): List<ItemInfo> {
        val itemCount = items.size
        val currentItemsInfo = ArrayList<ItemInfo>(itemCount)
        for (i in 0 until itemCount) {
            currentItemsInfo.add(
                    ItemInfo(getItemStringId(i),
                            getItemHash(i))
            )
        }
        return currentItemsInfo
    }

    /**
     * Automatically call necessary notify... methods.
     */
    private fun diffNotifyDataSetChanged() {
        val newItemInfo = extractRealItemInfo()
        val diffResult = DiffUtil.calculateDiff(
                AutoNotifyDiffCallback(lastItemsInfo,
                        newItemInfo,
                        allowInfiniteScroll)
        )
        diffResult.dispatchUpdatesTo(this)
        lastItemsInfo = newItemInfo
    }

    abstract class ViewHolder<I>(view: View) : RecyclerView.ViewHolder(view) {

        constructor(parent: ViewGroup, @LayoutRes layoutId: Int) :
                this(LayoutInflater.from(parent.context).inflate(layoutId, parent, false))

        protected val context: Context = view.context

        open fun bindData(position: Int, item: I, count: Int) {
            itemView.visibility = View.VISIBLE
        }

        open fun bindEmptyData(position: Int, item: I?, count: Int) {
            itemView.visibility = View.GONE
        }

        open fun onViewRecycled() {
            // override if needed
        }
    }

    protected class ItemsEventsObservable<I, VH : ViewHolder<*>> : Observable<ItemsEventsListener<I>>() {

        fun notifyItemClick(position: Int, item: I?) {
            synchronized(mObservers) {
                for (l in mObservers) {
                    l.onItemClick(position, item)
                }
            }
        }

        fun notifyItemLongClick(position: Int, item: I?): Boolean {
            synchronized(mObservers) {
                var consumed = false
                for (l in mObservers) {
                    if (l.onItemLongClick(position, item)) {
                        consumed = true
                    }
                }
                return consumed
            }
        }

        fun notifyItemFocused(position: Int, item: I?) {
            synchronized(mObservers) {
                for (l in mObservers) {
                    l.onItemFocusChanged(position, item)
                }
            }
        }

        fun notifyItemAdded(to: Int, item: I?) {
            synchronized(mObservers) {
                for (l in mObservers) {
                    l.onItemAdded(to, item)
                }
            }
        }

        fun notifyItemsAdded(to: Int, items: Collection<I>) {
            synchronized(mObservers) {
                for (l in mObservers) {
                    l.onItemsAdded(to, items)
                }
            }
        }

        fun notifyItemSet(to: Int, item: I?) {
            synchronized(mObservers) {
                for (l in mObservers) {
                    l.onItemSet(to, item)
                }
            }
        }

        fun notifyItemsSet(items: List<I?>) {
            synchronized(mObservers) {
                for (l in mObservers) {
                    l.onItemsSet(items)
                }
            }
        }

        fun notifyItemRemoved(from: Int, item: I?) {
            synchronized(mObservers) {
                for (l in mObservers) {
                    l.onItemRemoved(from, item)
                }
            }
        }

        fun notifyItemsRangeRemoved(from: Int, to: Int, previousSize: Int) {
            synchronized(mObservers) {
                for (l in mObservers) {
                    l.onItemsRangeRemoved(from, to, previousSize)
                }
            }
        }
    }

    interface ItemsEventsListener<I> {

        fun onItemClick(position: Int, item: I?)

        /**
         * @return true if event consumed
         */
        fun onItemLongClick(position: Int, item: I?): Boolean

        /**
         * Колбек при получении фокуса на указанную позицию
         */
        fun onItemFocusChanged(position: Int, item: I?)

        fun onItemAdded(to: Int, item: I?)

        fun onItemsAdded(to: Int, items: Collection<I>)

        fun onItemSet(to: Int, item: I?)

        fun onItemsSet(items: List<I?>)

        fun onItemRemoved(from: Int, item: I?)

        fun onItemsRangeRemoved(from: Int, to: Int, previousSize: Int)
    }
}
